docs_src/design/adr/security/0009-Secure-Bootstrapping.md

```diff
@@ -0,0 +1,492 @@
+and Consul itself is not configured in a secure mode.
+Technically speaking, Consul provides both a service registry
+and an HTTP-accessible configuration service--
+but because EdgeX uses configuration parameters
+to configure the host and port of outbound connections,
+the net effect is that Consul is used for service location.
+We need a bootstrapping mechanism that ensures that all services,
+including Conul itself, are brought up in a secure configuration.
+It is impossible,
+once certain services have been started (such as Consul),
+to go back and secure them on-the-fly.
+
+Secure bootstrapping is needed for any installation of EdgeX,
+but the problem is more profound for a container-based deployment of EdgeX
+because a docker-compose script does not have an explicit installation phase.
+The information in this ADR is generally applicable to the installation phase of an EdgeX snap,
+or any manual installation of EdgeX.
+
+A security-disabled installation of EdgeX is not required to follow this guidance.
+
+### Items requiring secure bootstrapping
+
+#### Secrets seed
+
+The secrets seed is a hardware-protected secret value that is used for key derivation.
+The **seed** can be used as input key material to a key deriviation function
+that generates an unlimited number of derived keys from an initial secret,
+all of which appear to be random.
+
+Derived keys can be used to protect bootstrapping secrets
+that cannot otherwise be protected by the EdgeX secret store.
+
+#### Consul
+
+Consul is a dynamic configuration database that is initially seeded from static configuration files.
+Putting configuration "online" exposes it to unauthorized modification and disclosure
+and requires introducing compensating controls.
+Most at risk is service location:
+information regarding the host and port where an EdgeX service is available.
+To a lesser extent,
+Consul holds general service configuration information
+and is the fallback location for the EdgeX secret store if Vault is not used.
+Consul also provides for service health checks.
+
+The elements required to secure Consul are:
+
+- For neighbor discovery: An **encryption key** used by Serf;
+  this key must be provided the first time that Consul is started.
+  This key prevents rogue Consul instances from joining the cluster.
+  This key must also be provided to any new Consul agents that are added to the system.
+  This key can be a 32-byte derived key or a random key.
+- For TLS: a **CA certificate**, **TLS certificate (chain)**, and **TLS private key**.
+
+For integration with Vault,
+some initial secrets must be created:
+
+- A **management token** that can be passed to Vault's Consul secrets engine,
+  to enable a Vault API to be utilized to create additional per-microservice Consul tokens.
+- An **ACL token** that allows Vault to read/write a portion of Consul's key space as the backing store for Vault.
+  (Intent is to switch to the Vault raft storage backend to avoid Consul dependency.)
```

Comment:
If this is still the intent, i.e `raft storage`, I think the integration section can just say there is no integration once Vault is switch to using `raft storage`, or something like that. i.e. no need to talk about Vault integration details that are needed.

Reply:
I will just delete *ACL token* -- a management token is still needed in order to have a single interface for generating Consul authentication tokens.

Reply:
Closing this comment as this section has been removed in the trimmed-down ADR.